game-core/src/main/java/games/strategy/triplea/ui/MovePanel.java

```diff
@@ -1123,119 +1107,37 @@ private void updateUnitsThatCanMoveOnRoute(final Collection<Unit> units, final R
       return;
     }
     getMap().hideMouseCursor();
-    final Collection<Unit> transportsToLoad = getPossibleTransportsToLoad(units, route);
-    List<Unit> best = new ArrayList<>(units);
-    // if the player selects a land unit and other units then
-    // only consider the non land units
-    if (route.getStart().isWater() && route.getEnd().isWater() && !route.isLoad()) {
-      best = CollectionUtils.getMatches(best, Matches.unitIsLand().negate());
-    }
-    if (route.isUnload()) {
-      best = CollectionUtils.getMatches(best, Matches.unitIsNotSea());
-    }
-    sortUnitsToMoveReverse(best, route);
-    MoveValidationResultWithDependents lastResult =
-        validateMoveWithDependents(best, transportsToLoad, route);
-    final MoveValidationResult allUnitsResult = lastResult.getResult();
-
-    if (!allUnitsResult.isMoveValid()) {
-      // if the player is invading only consider units that can invade
-      if (!nonCombat
-          && route.isUnload()
-          && Matches.isTerritoryEnemy(getCurrentPlayer(), getData()).test(route.getEnd())) {
-        lastResult = validateMoveWithDependents(best, transportsToLoad, route);
-      }
-      while (!best.isEmpty() && !lastResult.getResult().isMoveValid()) {
-        best = best.subList(1, best.size());
-        lastResult = validateMoveWithDependents(best, transportsToLoad, route);
-      }
-    }
-    if (allUnitsResult.isMoveValid()) {
-      // valid move
-      if (lastResult.getUnitsWithDependents().containsAll(selectedUnits)) {
-        clearStatusMessage();
-        currentCursorImage = null;
-      } else {
-        setStatusWarningMessage("Not all units can move there");
-        currentCursorImage = warningImage;
-      }
-    } else {
-      String message = allUnitsResult.getError();
-      if (message == null) {
-        message = allUnitsResult.getDisallowedUnitWarning(0);
-      }
-      if (message == null) {
-        message = allUnitsResult.getUnresolvedUnitWarning(0);
-      }
-      if (!lastResult.getResult().isMoveValid()) {
-        setStatusErrorMessage(message);
+
+    final MovableUnitsFilter unitsFilter =
+        new MovableUnitsFilter(getUnitOwner(units), route, nonCombat, moveType, getUndoableMoves());
+    final var result = unitsFilter.filterUnitsThatCanMove(units, dependentUnits);
+    switch (result.getStatus()) {
+      case NO_UNITS_CAN_MOVE:
+        setStatusErrorMessage(result.getWarningOrErrorMessage().orElseThrow());
```

Comment:
Just to convey, I was thinking the 'getWarningOrErrorMessage' String would be defined & set here. In that case the 'result' object would be a pure result without any strings in it. NBD.